nanofi/src/core/string_utils.c

```diff
@@ -0,0 +1,87 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "core/cstructs.h"
+#include "core/string_utils.h"
+#include <string.h>
+#include <stdlib.h>
+
+tokens * tokenize_string(char * str, const char * delimiter) {
```

Comment:
In my opinion this function works unexpectedly if:
 - there is no delimiter in the string (I think it should return with a single token containing the entire string)
 - for the last token after the last delimiter, before the terminating '\0': it does not create a token for it

I assume that it works this way because of the business logic we want to achieve in TailFile (namely not creating a token until we encounter a delimiter), but I don't think this function works the way most people would expect.